Control who gets notified by GerritStatusPush

dinatale2 opened the pull request:
Allow the GerritStatusPush object to accept who to notify as an argument. The notify argument is a string
which accepts values found in the gerrit documentation for the `--notify` flag. The argument is set to None by
default to maintain current functionality.

Signed-off-by: Giuseppe Di Natale <dinatale2@llnl.gov>

tardyp commented:
Hi we would appreciate some unit tests to test this feature, so that it does not reduce the coverage

sa2ajj commented on master/buildbot/reporters/gerrit.py:
```diff
@@ -175,6 +176,7 @@ def reconfigService(self, server, username, reviewCB=DEFAULT_REVIEW,
         self.summaryCB = summaryCB
         self.summaryArg = summaryArg
         self.builders = builders
+        self.gerrit_notify = notify
```
while legacy code extensively uses public members, the new code is expected to be aware whether it's a part of the class' API or not.  I'd suggest to use `_gerrit_notify`

sa2ajj commented on master/buildbot/reporters/gerrit.py:
```diff
@@ -397,6 +399,10 @@ def sendCodeReview(self, project, revision, result):
         command = self._gerritCmd("review", "--project %s" % (project,))
+
+        if self.gerrit_notify is not None:
+            command.extend(["--notify %s" % str(self.gerrit_notify)])
```
why do you use `.extend` for a single item?  use `.append`

please use `'` not `"`

sa2ajj commented on master/buildbot/test/unit/test_reporter_gerrit.py:
```diff
@@ -142,7 +142,8 @@ def setUp(self):
         serv = kwargs.pop("server", "serv")
         username = kwargs.pop("username", "user")
-        gsp = GerritStatusPush(serv, username, *args, **kwargs)
+        notify = kwargs.pop("notify", None)
```
ditto for `"`

sa2ajj commented on master/buildbot/test/unit/test_reporter_gerrit.py:
```diff
@@ -463,3 +464,22 @@ def testBuildGerritCommand(self):
+
+        # now test the notify argument
+        gsp.gerrit_notify = 'OWNER'
+        gsp.processVersion("2.6", lambda: None)
```
please make sure that all strings that do not include `'` use `"`

dinatale2 replied:
do you mind clarifying when to use `'` instead of `"`? For one of the other comments above (in the .extend comment), you suggested I use `'` instead of `"` is why I ask.

sa2ajj replied:
Well, per [String Quotes](https://www.python.org/dev/peps/pep-0008/#string-quotes) in PEP-8, both `"` and `'` are OK, however it's good to stick to some particular way.

So far, we've been using `"` for docstrings only `"""` and `'` for all other string literals (some legacy code does not follow this, I must admit).

sa2ajj commented on master/buildbot/test/unit/test_reporter_gerrit.py:
```diff
@@ -463,3 +463,23 @@ def testBuildGerritCommand(self):
+
+        # now test the notify argument, even though _gerrit_notify
+        # is private, work around that
+        gsp._gerrit_notify = "OWNER"
```
please make sure that you use single quotes everywhere (even in tests).

sa2ajj commented on master/buildbot/test/unit/test_reporter_gerrit.py:
```diff
@@ -463,3 +463,23 @@ def testBuildGerritCommand(self):
+            ['ssh', 'user@serv', '-p', '29418', 'gerrit', 'review',
+             '--project project', '--notify OWNER', "--message 'bla'", '--label Verified=1', 'revision'])
+
+        # <=2.5 uses other syntax
```
2.5 of what?

dinatale2 commented:
I've updated this replacing the `"` with `'` and I've introduced a comment about the different syntax with gerrit versions 2.5 and below. I'm not sure why the bb test failed as well, it appears to have been cancelled.

tardyp commented:
looks good. There is stil a need for documentation update, and release notes

dinatale2 commented:
docs and release notes have been updated. Let me know if I need to make any additional changes.

tardyp commented on master/docs/manual/cfg-reporters.rst:
```diff
@@ -738,6 +738,7 @@ GerritStatusPush can send a separate review for each build that completes, or a
    :param builders: (optional) list of builders to send results for.
                     This method allows to filter results for a specific set of builder.
                     By default, or if builders is None, then no filtering is performed.
+   :param notify: (optional) control who gets notified by Gerrit once the status is posted.
```
I think this would deserve to document which values are possible, probably point on the gerrit documentation, explaining that this correspond to the --notify values.

dinatale2 replied:
Updated.
